### ISSUE TITLE
Visualization results

### ISSUE THREAD

**Issue by fufj** [ruili3/dynamic-multiframe-depth#16]
Hi~ Thanks for your remarkable work. 
I am reproducing your paper results including quantitative and qualitative results. However, difficulties are encountered in generating the visualization results as shown in Figure 6 of your paper. I have found taht two .py files in this repository, plot_depth_utils.py and visualization.py, may be related. 
Would you mind providing some help for me?

**Comment by ruili3** [comment]
Hi,

Thanks for your interest :)

- We used the AbsRel error map in Fig.6, you may alter the `save_concat_res()` function in `plot_depth_utils.py` for error visualization.
- Unfortunately, the current codebase does not contain the original point cloud generation script. Our baseline work has a point cloud generation [script](https://github.com/Brummi/MonoRec/blob/81b8dd98b86e590069bb85e0c980a03e7ad4655a/create_pointcloud.py#L16), but it requires further debugging to work out. An alternative is to use other repos, eg. these functions ([pcd](https://github.com/YvanYin/Metric3D/blob/2118ec93a24f0f80204f56be96459ac049d37bf4/mono/utils/unproj_pcd.py#L25), [save_pcd](https://github.com/YvanYin/Metric3D/blob/2118ec93a24f0f80204f56be96459ac049d37bf4/mono/utils/unproj_pcd.py#L38)) in Metric3D. It should work fine in our repo if you set the intrinsic properly.

**Comment by fufj** [comment]
Thank you so much. I will try it.

**Comment by 757787182** [comment]
![error_map](https://github.com/user-attachments/assets/4f6221a7-a0da-4d4f-8063-82156bc8a13c)
Hi~ Thanks for your remarkable work.
my loss of error chart looks like this,Would you mind providing some help for me?

**Comment by ruili3** [comment]
> ![error_map](https://private-user-images.githubusercontent.com/83005849/352488750-4f6221a7-a0da-4d4f-8063-82156bc8a13c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE5OTAxODIsIm5iZiI6MTcyMTk4OTg4MiwicGF0aCI6Ii84MzAwNTg0OS8zNTI0ODg3NTAtNGY2MjIxYTctYTBkYS00ZDRmLTgwNjMtODIxNTZiYzhhMTNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI2VDEwMzEyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdiNTFhZTc3ZWI4YjkxYTk2ODA2MDY2ZjcxYTk0NzgwM2FmMzE1ZDdiMjI5MDg0MTQ4MzdlNjA3NDVlMzdiMzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.okn4ntvnpJ6h7nrHFfgXqz2-myFO27mSQWz2AahuZwA) Hi~ Thanks for your remarkable work. my loss of error chart looks like this,Would you mind providing some help for me?

Did you use the script I mentioned above? It seems that you are using the original image and a different color scheme

**Comment by 757787182** [comment]
thanks you for your reply ,i will try it

**Comment by xuhang2017** [comment]
Hello, I am very interested in your work, but I did not find the visualization code of the depth image in your code, it is convenient for you to provide it, thank you

**Comment by ruili3** [comment]
> Hello, I am very interested in your work, but I did not find the visualization code of the depth image in your code, it is convenient for you to provide it, thank you

Hi, thanks for your interest. Have you tried using the following code as I mentioned above?

> We used the AbsRel error map in Fig.6, you may alter the save_concat_res() function in plot_depth_utils.py for error visualization.
Unfortunately, the current codebase does not contain the original point cloud generation script. Our baseline work has a point cloud generation [script](https://github.com/Brummi/MonoRec/blob/81b8dd98b86e590069bb85e0c980a03e7ad4655a/create_pointcloud.py#L16), but it requires further debugging to work out. An alternative is to use other repos, eg. these functions ([pcd](https://github.com/YvanYin/Metric3D/blob/2118ec93a24f0f80204f56be96459ac049d37bf4/mono/utils/unproj_pcd.py#L25), [save_pcd](https://github.com/YvanYin/Metric3D/blob/2118ec93a24f0f80204f56be96459ac049d37bf4/mono/utils/unproj_pcd.py#L38)) in Metric3D. It should work fine in our repo if you set the intrinsic properly.

Feel free if you encounter any difficulties

**Comment by xuhang2017** [comment]
> > Hello, I am very interested in your work, but I did not find the visualization code of the depth image in your code, it is convenient for you to provide it, thank you
> 
> Hi, thanks for your interest. Have you tried using the following code as I mentioned above?
> 
> > We used the AbsRel error map in Fig.6, you may alter the save_concat_res() function in plot_depth_utils.py for error visualization.
> > Unfortunately, the current codebase does not contain the original point cloud generation script. Our baseline work has a point cloud generation [script](https://github.com/Brummi/MonoRec/blob/81b8dd98b86e590069bb85e0c980a03e7ad4655a/create_pointcloud.py#L16), but it requires further debugging to work out. An alternative is to use other repos, eg. these functions ([pcd](https://github.com/YvanYin/Metric3D/blob/2118ec93a24f0f80204f56be96459ac049d37bf4/mono/utils/unproj_pcd.py#L25), [save_pcd](https://github.com/YvanYin/Metric3D/blob/2118ec93a24f0f80204f56be96459ac049d37bf4/mono/utils/unproj_pcd.py#L38)) in Metric3D. It should work fine in our repo if you set the intrinsic properly.
> 
> Feel free if you encounter any difficulties

Thank you so much. I will try it.

**Comment by kehuitt** [comment]
Hi, thanks for your remarkable work!
I'm attempting to visualize the results using the script mentioned above. However, I can not obtain the true visualized outcomes. The relevant revised code is presented below. Would you identify what needs modification?
```python
# evaluate.py
for batch_idx, (data, target) in enumerate(self.data_loader):
    data, target = to(data, self.device), to(target, self.device)
    data["target"] = target

    with torch.no_grad():
        data = self.model(data)
        loss_dict = {"loss": torch.tensor([0])}
        loss = loss_dict["loss"]

    output = data["result"]

    pred_disp = output.cpu().numpy()
    gt_depth = target.cpu().numpy()
    img = data["keyframe"].cpu().numpy()
    self.save_visualization(batch_idx, pred_disp, img, gt_depth)

def save_visualization(self, batch_idx, pred_disp, img, gt_depth):
    """
    :param pred_disp: [batch_size, h, w]
    :param img: [batch_size, c, h, w]
    :param gt_depth: [batch_size, h, w]
    """

    output_dir  = "./visualization_results"
    if not os.path.exists(output_dir):
        os.mkdir(output_dir)

    for i in range(gt_depth.shape[0]):
        sample_pred_disp = np.transpose(pred_disp[i], (1, 2, 0))
        sample_gt_depth = np.transpose(gt_depth[i], (1, 2, 0))
        sample_img = np.transpose(img[i], (1, 2, 0))
        save_concat_res(output_dir, sample_pred_disp, sample_img, sample_gt_depth)
```

**Comment by ruili3** [comment]
> Hi, thanks for your remarkable work! I'm attempting to visualize the results using the script mentioned above. However, I can not obtain the true visualized outcomes. The relevant revised code is presented below. Would you identify what needs modification?
> 
> # evaluate.py
> for batch_idx, (data, target) in enumerate(self.data_loader):
>     data, target = to(data, self.device), to(target, self.device)
>     data["target"] = target
> 
>     with torch.no_grad():
>         data = self.model(data)
>         loss_dict = {"loss": torch.tensor([0])}
>         loss = loss_dict["loss"]
> 
>     output = data["result"]
> 
>     pred_disp = output.cpu().numpy()
>     gt_depth = target.cpu().numpy()
>     img = data["keyframe"].cpu().numpy()
>     self.save_visualization(batch_idx, pred_disp, img, gt_depth)
> 
> def save_visualization(self, batch_idx, pred_disp, img, gt_depth):
>     """
>     :param pred_disp: [batch_size, h, w]
>     :param img: [batch_size, c, h, w]
>     :param gt_depth: [batch_size, h, w]
>     """
> 
>     output_dir  = "./visualization_results"
>     if not os.path.exists(output_dir):
>         os.mkdir(output_dir)
> 
>     for i in range(gt_depth.shape[0]):
>         sample_pred_disp = np.transpose(pred_disp[i], (1, 2, 0))
>         sample_gt_depth = np.transpose(gt_depth[i], (1, 2, 0))
>         sample_img = np.transpose(img[i], (1, 2, 0))
>         save_concat_res(output_dir, sample_pred_disp, sample_img, sample_gt_depth)

Hi, could you please specify what issues specifically?

**Comment by kehuitt** [comment]
@ruili3 Hi, I utilized the `save_concat_res(output_dir, sample_pred_disp, sample_img, sample_gt_depth)` function from `plot_depth_utils.py`, and made revisions to the code in `evaluate.py`. Here is the result I obtained. (I have not modified any code in `plot_depth_utils.py`)
Except for the second picture, the other three pictures seem to be wrong, is there something wrong with the code I modified in `eval.py`? Thanks!!

![Image](https://github.com/user-attachments/assets/c7f791b1-be1d-4fcf-a569-f812c7e4c31a)

**Comment by ruili3** [comment]
> [@ruili3](https://github.com/ruili3) Hi, I utilized the `save_concat_res(output_dir, sample_pred_disp, sample_img, sample_gt_depth)` function from `plot_depth_utils.py`, and made revisions to the code in `evaluate.py`. Here is the result I obtained. (I have not modified any code in `plot_depth_utils.py`) Except for the second picture, the other three pictures seem to be wrong, is there something wrong with the code I modified in `eval.py`? Thanks!!
> 
> ![Image](https://github.com/user-attachments/assets/c7f791b1-be1d-4fcf-a569-f812c7e4c31a)

thanks for the feedback! it should work fine with the following changes:

- adding 0.5 to the input image and clipping the value between [0,1]: the "keyframe" was originally mapped to [-0.5, 0.5]
- transferring GT disp into metric depth following the same way as was done to the prediction: https://github.com/ruili3/dynamic-multiframe-depth/blob/b3d3f4eb00052c9d1f42c5f0abfaf896fceeb39e/depth_proc_tools/plot_depth_utils.py#L187C7-L187C85

feel free to let me know if further assistance is needed!
